File: Website/js/beam.js
class Beam {
    constructor(joint1Idx, joint2Idx) {
        this.joint1Idx = joint1Idx;
        this.joint2Idx = joint2Idx;
        this.force = 0.0;
        this.crossSectionalArea = BEAM_DEFAULT_AREA;
        this.stress = 0.0;
    }

    draw(ctx, joints, worldToScreenFunc, showForces) {
        if (this.joint1Idx >= joints.length || this.joint2Idx >= joints.length) {
            return;
        }

        const joint1 = joints[this.joint1Idx];
        const joint2 = joints[this.joint2Idx];
        const [screenX1, screenY1] = worldToScreenFunc([joint1.x, joint1.y]);
        const [screenX2, screenY2] = worldToScreenFunc([joint2.x, joint2.y]);

        let color = BEAM_COLOR;
        
        // If showing forces and beam has a force, use a color gradient based on tension/compression
        if (showForces && Math.abs(this.force) > FLOAT_TOLERANCE) {
            const ratio = Math.min(Math.abs(this.force) / MAX_FORCE_VIS, 1.0);
            if (this.force > 0) {
                // Tension - show in red
                const intensity = Math.floor(255 * (1 - ratio));
                color = `rgb(255, ${intensity}, ${intensity})`;
            } else {
                // Compression - show in blue
                const intensity = Math.floor(255 * (1 - ratio));
                color = `rgb(${intensity}, ${intensity}, 255)`;
            }
        }

        // Draw the beam
        ctx.beginPath();
        ctx.moveTo(screenX1, screenY1);
        ctx.lineTo(screenX2, screenY2);
        ctx.strokeStyle = color;
        ctx.lineWidth = 5;
        ctx.stroke();
    }

    getLength(joints) {
        if (this.joint1Idx >= joints.length || this.joint2Idx >= joints.length) {
            return 0;
        }
        
        const joint1 = joints[this.joint1Idx];
        const joint2 = joints[this.joint2Idx];
        
        return Math.hypot(joint1.x - joint2.x, joint1.y - joint2.y);
    }

    getAngleRadRelativeTo(jointIdx, joints) {
        if (this.joint1Idx >= joints.length || this.joint2Idx >= joints.length) {
            return null;
        }
        
        let startJoint, endJoint;
        
        if (this.joint1Idx === jointIdx) {
            startJoint = joints[this.joint1Idx];
            endJoint = joints[this.joint2Idx];
        } else if (this.joint2Idx === jointIdx) {
            startJoint = joints[this.joint2Idx];
            endJoint = joints[this.joint1Idx];
        } else {
            return null;
        }
        
        return Math.atan2(endJoint.y - startJoint.y, endJoint.x - startJoint.x);
    }
} 

File: simulation/beam.py
import pygame
import math
from config import *

class Beam:
    def __init__(self, joint1_idx, joint2_idx):
        self.joint1_idx = joint1_idx
        self.joint2_idx = joint2_idx
        self.force = 0.0
        self.cross_sectional_area = BEAM_DEFAULT_AREA
        self.stress = 0.0

    def draw(self, screen, joints, world_to_screen_func, show_forces):
        if self.joint1_idx >= len(joints) or self.joint2_idx >= len(joints):
            return

        joint1 = joints[self.joint1_idx]
        joint2 = joints[self.joint2_idx]
        screen_joint1 = world_to_screen_func((joint1.x, joint1.y))
        screen_joint2 = world_to_screen_func((joint2.x, joint2.y))

        if show_forces and abs(self.force) > FLOAT_TOLERANCE:
            ratio = min(abs(self.force) / MAX_FORCE_VIS, 1.0)
            color = (255, int(255 * (1 - ratio)), int(255 * (1 - ratio))) if self.force > 0 else (int(255 * (1 - ratio)), int(255 * (1 - ratio)), 255)
        else:
            color = BEAM_COLOR

        pygame.draw.line(screen, color, screen_joint1, screen_joint2, 5)

    def get_length(self, joints):
        if self.joint1_idx >= len(joints) or self.joint2_idx >= len(joints):
            return 0
        joint1 = joints[self.joint1_idx]
        joint2 = joints[self.joint2_idx]
        return math.hypot(joint1.x - joint2.x, joint1.y - joint2.y)

    def get_angle_rad_relative_to(self, joint_idx, joints):
        if self.joint1_idx >= len(joints) or self.joint2_idx >= len(joints):
            return None
        if self.joint1_idx == joint_idx:
            start_joint, end_joint = joints[self.joint1_idx], joints[self.joint2_idx]
        elif self.joint2_idx == joint_idx:
            start_joint, end_joint = joints[self.joint2_idx], joints[self.joint1_idx]
        else:
            return None
        return math.atan2(end_joint.y - start_joint.y, end_joint.x - start_joint.x)